src/economy.cpp

```diff
@@ -1034,38 +1034,55 @@ static SmallIndustryList _cargo_delivery_destinations;
  */
 static uint DeliverGoodsToIndustry(const Station *st, CargoID cargo_type, uint num_pieces, IndustryID source)
 {
-	/* Find the nearest industrytile to the station sign inside the catchment area, whose industry accepts the cargo.
+	/* Find a random industry near the station sign inside the catchment area, whose industry accepts the cargo.
 	 * This fails in three cases:
 	 *  1) The station accepts the cargo because there are enough houses around it accepting the cargo.
 	 *  2) The industries in the catchment area temporarily reject the cargo, and the daily station loop has not yet updated station acceptance.
 	 *  3) The results of callbacks CBID_INDUSTRY_REFUSE_CARGO and CBID_INDTILE_CARGO_ACCEPTANCE are inconsistent. (documented behaviour)
 	 */
 
 	uint accepted = 0;
+	if (st->industries_near.Length() == 0) return accepted;
 
-	for (uint i = 0; i < st->industries_near.Length() && num_pieces != 0; i++) {
-		Industry *ind = st->industries_near[i];
-		if (ind->index == source) continue;
-
+	SmallIndustryList rejected;
+	do {
+		Industry *ind = st->industries_near[RandomRange(st->industries_near.Length())];
+		if (ind->index == source) {
+			rejected.Include(ind);
+			continue;
+		}
```

Comment:
There's a lot of repetition of this part. If you extract all the tests to a separate function, you can reduce the repetition.